Sources/NIO/ConvenienceOptionSupport.swift

```diff
@@ -0,0 +1,178 @@
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// MARK:  Universal Client Bootstrap
+extension NIOClientTCPBootstrapProtocol {
+    /// Apply any understood shorthand options to the bootstrap, removing them from the set of options if they are consumed.
+    /// - parameters:
+    ///     - options:  The options to try applying - the options applied should be consumed from here.
+    /// - returns: The updated bootstrap with and options applied.
+    public func _applyOptions(_ options: inout NIOTCPShorthandOptions) -> Self {
+        // Default is to consume no options and not update self.
+        return self
+    }
+}
+
+extension NIOClientTCPBootstrap {
+    /// Specifies some `ChannelOption`s to be applied to the channel.
+    /// - See: channelOption
+    /// - Parameter options: Set of shorthand options to apply.
+    /// - Returns: The updated bootstrap (`self` being mutated)
+    public func options(_ options: NIOTCPShorthandOptions) -> NIOClientTCPBootstrap {
+        var optionsRemaining = options
+        // First give the underlying a chance to consume options.
+        let withUnderlyingOverrides =
+            NIOClientTCPBootstrap(self, withUpdated: underlyingBootstrap._applyOptions(&optionsRemaining))
+        // Default apply any remaining options.
+        return optionsRemaining.applyFallbackMapping(withUnderlyingOverrides)
+    }
+}
+
+// MARK: Utility
+/// Has an option been set?
+/// Option has a value of generic type T.
+public enum NIOOptionValue<T> {
+    /// The option was not set.
+    case notSet
+    /// The option was set with a value of type T.
+    case set(T)
+}
+
+public extension NIOOptionValue where T == () {
+    /// Convenience method working with bool options as bool values for set.
+    var isSet: Bool {
+        get {
+            switch self {
+            case .notSet:
+                return false
+            case .set(()):
+                return true
+            }
+        }
+    }
+}
+
+private extension NIOOptionValue where T == () {
+    init(flag: Bool) {
+        if flag {
+            self = .set(())
+        } else {
+            self = .notSet
+        }
+    }
+}
+
+// MARK: TCP - data
+/// A TCP channel option which can be applied to a bootstrap using shorthand notation.
+public struct NIOTCPShorthandOption: Hashable {
+    fileprivate var data: ShorthandOption
+    
+    private init(_ data: ShorthandOption) {
+        self.data = data
+    }
+    
+    fileprivate enum ShorthandOption: Hashable {
+        case reuseAddr
+        case disableAutoRead
+        case allowRemoteHalfClosure
+    }
+}
+
+/// Approved shorthand options.
+extension NIOTCPShorthandOption {
+    /// Allow immediately reusing a local address.
+    public static let allowImmediateLocalEndpointAddressReuse = NIOTCPShorthandOption(.reuseAddr)
```

Comment:
@Davidde94 / @glbrntt / @Lukasa these three names need bike-shedding.

Reply:
Not sure I'm a fan of `endpoint` being in the name, I think the others are fine IMO

Reply:
Maybe invert `disableAutoRead` to `enableAutoRead`. It feels a bit weird to have "allow", "allow", "don't allow"

Reply:
I think endpoint and address are redundant, pick one. Network.framework uses `Endpoint` which may be prior art justifying that name.

Reply:
Given that `autoRead` is on by default, I don't think we can have an option to enable it without changing that default, which we cannot safely do in a semver minor.

Reply:
Agree - ship has sailed autoRead defaults.
I think I'll go with Address - I like it more than Endpoint - `sockaddr_in` feels like prior art for it :-)